test/framework/multi_node_control_plane.go

```diff
@@ -0,0 +1,128 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package framework
+
+import (
+	"context"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// MultiNodeControlplaneClusterInput defines the necessary dependencies to run a multi-node control plane cluster.
+type MultiNodeControlplaneClusterInput struct {
+	Management        ManagementCluster
+	Cluster           *clusterv1.Cluster
+	InfraCluster      runtime.Object
+	ControlplaneNodes []Node
+	CreateTimeout     time.Duration
+}
+
+// SetDefaults defaults the struct fields if necessary.
+func (m *MultiNodeControlplaneClusterInput) SetDefaults() {
+	if m.CreateTimeout == 0 {
+		m.CreateTimeout = 10 * time.Minute
+	}
+}
+
+// MulitNodeControlPlaneCluster creates an n node control plane cluster.
+// Assertions:
+//  * The number of nodes in the created cluster will equal the number of nodes in the input data.
+func MultiNodeControlPlaneCluster(input *MultiNodeControlplaneClusterInput) {
```

Comment:
This looks good for the happy path, but I'm worried about the state things will be left it if testing fails. Maybe some methods to pull logs out of the cluster would be helpful.

Reply:
I've implemented something along these lines in capa here: https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/651566d34bb7ca79b90e1dc2e835c7333d7c1f80/test/e2e/e2e_suite_test.go#L204

Reply:
Watching events helps with troubleshooting as well: https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/651566d34bb7ca79b90e1dc2e835c7333d7c1f80/test/e2e/aws_test.go#L158

Reply:
For sure agreed. I'm going to add this to the main tracking issue since this can be designed and worked on independently, probably heavily inspired by the links jason has pasted